helix-view/src/editor.rs

```diff
@@ -662,6 +670,19 @@ impl Editor {
             .find(|doc| doc.path().map(|p| p == path.as_ref()).unwrap_or(false))
     }
 
+    /// Get the document's auto pairs. If the document has a recognized
+    /// language config with auto pairs configured, returns that;
+    /// otherwise, falls back to the global auto pairs config.
+    pub fn get_document_auto_pairs(&self, id: DocumentId) -> Option<&AutoPairs> {
+        let global_config = (&self.auto_pairs).as_ref();
+        let doc = self.document(id)?;
```

Comment:
Hmm, I see what you mean, but I don't think we'd want to pass in the editor config, because then we'd be reconstructing a `HashMap` on every keystroke. And if we put the parsed `AutoPairs` object in the `Document` instead, then we'd have a choice between:

* Each `Document` would need to have a copy of the parsed `AutoPairs`, which is recreated from the config on every new `Document` or
* Still keep the parsed `AutoPairs` at the `Editor` level and have an `Rc` member in the `Document`

Neither would be an awful choice, but it seems like a lot just to avoid an array index operation to look up the same `Document`.

Reply:
I'm not following, why would we be reconstructing the map? It would be the exact same function, but `doc` would be the `&self` receiver, and `&editor`/`&editor.config` as the passed in global_config.

Even without the change to the API, the function signature could be `pub fn get_document_auto_pairs(&'_ self, doc: &Document) -> Option<&'_ AutoPairs>`

Reply:
> I'm not following, why would we be reconstructing the map? It would be the exact same function, but `doc` would be the `&self` receiver, and `&editor`/`&editor.config` as the passed in global_config.

Oh, in your earlier comment, you suggested passing in `editor.config`, but this is just the representation of the config, not the parsed auto pairs. If all we have is the config, then we'd have to get the parsed representation of it somehow, so we'd have to either parse it again or keep a reference to the parsed object, like I mentioned above.

However, I hadn't considered using the whole `Editor`. I changed it to a method on `Document` with this approach.